src/main/java/com/google/api/generator/engine/ast/RelationalOperationExpr.java

```diff
@@ -0,0 +1,135 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.engine.ast;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+
+@AutoValue
+public abstract class RelationalOperationExpr implements OperationExpr {
+  public abstract Expr lhsExpr();
+
+  public abstract Expr rhsExpr();
+
+  public abstract OperatorKind operatorKind();
+
+  @Override
+  public TypeNode type() {
+    return TypeNode.BOOLEAN;
+  }
+
+  @Override
+  public void accept(AstNodeVisitor visitor) {
+    visitor.visit(this);
+  }
+
+  // Convenience wrapper.
+  public static RelationalOperationExpr equalToWithExprs(Expr lhsExpr, Expr rhsExpr) {
+    return builder()
+        .setLhsExpr(lhsExpr)
+        .setRhsExpr(rhsExpr)
+        .setOperatorKind(OperatorKind.RELATIONAL_EQUAL_TO)
+        .build();
+  }
+
+  // Convenience wrapper.
+  public static RelationalOperationExpr notEqualToWithExprs(Expr lhsExpr, Expr rhsExpr) {
+    return builder()
+        .setLhsExpr(lhsExpr)
+        .setRhsExpr(rhsExpr)
+        .setOperatorKind(OperatorKind.RELATIONAL_NOT_EQUAL_TO)
+        .build();
+  }
+
+  // TODO(summerji): Add convenience wrapper lessThanWithExprs
+  // public static RelationalOperationExpr lessThanWithExprs(Expr lhsExpr, Expr rhsExpr) {
+  //   return builder()
+  //       .setLhsExpr(lhsExpr)
+  //       .setRhsExpr(rhsExpr)
+  //       .setOperatorKind(OperatorKind.RELATIONAL_LESS_THAN)
+  //       .build();
+  // }
+
+  private static Builder builder() {
+    return new AutoValue_RelationalOperationExpr.Builder();
+  }
+
+  @AutoValue.Builder
+  abstract static class Builder {
+
+    // Private setter.
+    abstract Builder setLhsExpr(Expr expr);
+
+    // Private setter.
+    abstract Builder setRhsExpr(Expr expr);
+
+    // Private setter.
+    abstract Builder setOperatorKind(OperatorKind operator);
+
+    abstract RelationalOperationExpr autoBuild();
+
+    private RelationalOperationExpr build() {
+      RelationalOperationExpr relationalOperationExpr = autoBuild();
+      TypeNode lhsExprType = relationalOperationExpr.lhsExpr().type();
+      TypeNode rhsExprType = relationalOperationExpr.rhsExpr().type();
+      OperatorKind operator = relationalOperationExpr.operatorKind();
+      final String errorMsg =
+          String.format(
+              "Relational operator %s can not be applied to %s, %s.",
+              operator, lhsExprType.toString(), rhsExprType.toString());
+
+      if (operator.equals(OperatorKind.RELATIONAL_EQUAL_TO)
+          || operator.equals(OperatorKind.RELATIONAL_NOT_EQUAL_TO)) {
+        Preconditions.checkState(isValidEqualityType(lhsExprType, rhsExprType), errorMsg);
```

Comment:
Consider simplifying `isValidEqualityType` further to do only a one-way check (i.e. assume `lhs` and `rhs` are always in those  positions). Then, we can just call `isValidEqualityType(rhs, lhs)` to check the opposite condition.

Reply:
Can you clarify a bit more on the benefit of doing further simplification? Currently the equality is symetric, meaning if A = B is valid, then B = A will pass the compiler as well.

Reply:
It's great that the current logic is implemented in a symmetric way. Benefits of simplification include easier readability and maintainability. For example, someone reading the method doesn't need to check that the symmetry is implemented, and can be confident that symmetry will come from the second invocation. Not having to implement symmetry will also make it easier for someone to modify one-way logic or add checks.

Marking this as optional for a separate PR.

Reply:
Good discussion here. I created a ticket for refactoring type-checking.